Neos.Neos/Classes/AssetUsage/CatchUpHook/AssetUsageCatchUpHook.php

```diff
@@ -0,0 +1,185 @@
+use Neos\ContentRepository\Core\Feature\NodeModification\Event\NodePropertiesWereSet;
+use Neos\ContentRepository\Core\Feature\NodeRemoval\Event\NodeAggregateWasRemoved;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodeGeneralizationVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodePeerVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\NodeVariation\Event\NodeSpecializationVariantWasCreated;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Dto\NodeIdsToPublishOrDiscard;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasDiscarded;
+use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasPartiallyDiscarded;
+use Neos\ContentRepository\Core\Projection\CatchUpHookInterface;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\FindDescendantNodesFilter;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ContentGraph\VisibilityConstraints;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\EventStore\Model\EventEnvelope;
+use Neos\Neos\AssetUsage\Service\AssetUsageIndexingService;
+
+/**
+ * @internal
+ */
+class AssetUsageCatchUpHook implements CatchUpHookInterface
+{
+    private static bool $enabled = true;
+
+
+    public function __construct(
+        private readonly ContentRepository $contentRepository,
+        private readonly AssetUsageIndexingService $assetUsageIndexingService
+    ) {
+    }
+
+    public function canHandle(EventInterface $event): bool
+    {
+        return in_array($event::class, [
+            DimensionSpacePointWasMoved::class,
+            NodeAggregateWasRemoved::class,
+            NodeAggregateWithNodeWasCreated::class,
+            NodeGeneralizationVariantWasCreated::class,
+            NodeSpecializationVariantWasCreated::class,
+            NodePeerVariantWasCreated::class,
+            NodePropertiesWereSet::class,
+            WorkspaceWasDiscarded::class,
+            WorkspaceWasPartiallyDiscarded::class
+        ]);
+    }
+
+    public function onBeforeCatchUp(): void
+    {
+    }
+
+    public function onBeforeEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        if (!self::$enabled) {
+            // performance optimization: on full replay, we assume all caches to be flushed anyways
+            // - so we do not need to do it individually here.
+            return;
+        }
+
+        if (
+            $eventInstance instanceof EmbedsWorkspaceName
+            && $eventInstance instanceof EmbedsContentStreamId
+            && !$this->contentRepository->getWorkspaceFinder()->findOneByCurrentContentStreamId($eventInstance->getContentStreamId())?->workspaceName->equals($eventInstance->getWorkspaceName())
```

Comment:
i think we should prefer not using `findOneByCurrentContentStreamId` - also below.
```php
try {
    $contentGraph = $this->contentRepository->getContentGraph($eventInstance->getWorkspaceName());
    if (!$contentGraph->getContentStreamId()->equals($eventInstance->getContentStreamId()) {
        return;
    }
} catch (WorkspaceDoesNotExist) {
    return:
}
```

Reply:
try/catch for control flow is not great and costs performance

Reply:
but that is the bargain when we made `getContentGraph` throw instead of nullable ...

Reply:
@mhsdesign you're right, my comment wasn't really qualified.
And since we plan to remove the `WorkspaceFinder` with #5096 we should probably adjust this (we did a similar thing with #5270).

Reply:
Done

Reply:
Thanks exactly my thoughts :D